sandbox/samples/cpp/alembicassembly/alembicassembly.cpp

```diff
@@ -0,0 +1,833 @@
+        }
+
+        auto motion_segment_count = 0;
+        const auto num_samples = schema.getNumSamples();
+        std::cout << "num_samples " << num_samples << std::endl;
+
+        std::vector<Alembic::Abc::chrono_t> sample_times;
+
+        if (num_samples < 2)
+        {
+            std::cout << "No sampling " << std::endl;
+            sample_times.push_back(0.0f);
+        }
+        else  // more than 1 sample
+        {
+            // generate offsetted times to get samples
+            const auto t_open = m_shutter_open_time+m_time_offset;
+            const auto t_close = m_shutter_close_time+m_time_offset;
+
+            const auto time_sampling = schema.getTimeSampling();
+
+            // get floor and ceil sample ids to don't miss any
+            const auto open_pair = time_sampling->getFloorIndex(t_open, num_samples);
+            const auto close_pair = time_sampling->getCeilIndex(t_close, num_samples);
+
+            std::cout << "open_pair.first " << open_pair.first << std::endl;
+            std::cout << "close_pair.first " << close_pair.first << std::endl;
+
+            sample_times.reserve(close_pair.first-open_pair.first+1);
+
+            // get _every_ sample times between open and close
+            for (auto i = open_pair.first; i <= close_pair.first; i++)
+            {
+                const auto t = time_sampling->getSampleTime(i);
+                sample_times.push_back(t);
+            }
+
+            // compute appleseed number of motion segment
+            motion_segment_count = sample_times.size() - 1;
+
+            if(motion_segment_count)
+            {
+                if(!linearly_sampled(m_shutter_open_time, m_shutter_close_time, sample_times))
+                {
+                    std::cout << "not linearly sampled!" << std::endl;
+                    RENDERER_LOG_WARNING("Motion samples are not matching shutters.");  // TODO: logs doesn't work.
+                }
+            }
+        }
+
+        std::cout << "sample_times.size() " << sample_times.size() << std::endl;
+        for(const auto t : sample_times)
+        {
+            std::cout << "  t: " << t << " - frame: " << t * 24.0f << std::endl;
+        }
+
+        //getVelocities()
+
+        // Create a new mesh object
+        auto object = asf::auto_release_ptr<asr::MeshObject>(
+            asr::MeshObjectFactory().create(
+                name.c_str(),
+                asr::ParamArray()));
+
+        object->set_motion_segment_count(motion_segment_count);
+
+        ///////////////////////////////////////////////////////////////
+        // vertex positions
+        ///////////////////////////////////////////////////////////////
+        auto sample_id = 0;
+        for(const auto sample_time : sample_times)
+        {
+            const auto sample_sel = Alembic::AbcGeom::ISampleSelector(sample_time);
+
+            const auto sample = schema.getValue(sample_sel);
+
+            const auto pos = sample.getPositions()->get();
```

Comment:
What does `get()` return? Is this making an expensive copy? It's impossible to tell... I'd suggest not using `auto` when the type is not obvious.